native/dlaqr5.go

```diff
@@ -0,0 +1,609 @@
+// otherwise Dlaqr5 will panic.
+//
+// sr and si contain the real and imaginary parts, respectively, of the shifts
+// of origin that define the multi-shift QR sweep. On return both slices may be
+// reordered by Dlaqr5. Their length must be equal to nshfts, otherwise Dlaqr5
+// will panic.
+//
+// h and ldh represent the Hessenberg matrix H of size n×n. On return
+// multi-shift QR sweep with shifts sr+i*si has been applied to the isolated
+// diagonal block in rows and columns ktop through kbot, inclusive.
+//
+// iloz and ihiz specify the rows of Z to which transformations will be applied
+// if wantz is true. It must hold that 0 <= iloz <= ihiz < n, otherwise Dlaqr5
+// will panic.
+//
+// z and ldz represent the matrix Z of size n×n. If wantz is true, the QR sweep
+// orthogonal similarity transformation is accumulated into
+// z[iloz:ihiz,ilo:ihi] from the right, otherwise z not referenced.
+//
+// v and ldv represent an auxiliary matrix of size (nshfts/2)×3.
+//
+// u and ldv represent an auxiliary matrix of size (3*nshfts-3)×(3*nshfts-3).
+//
+// wh and ldwh represent an auxiliary matrix of size (3*nshfts-3)×nh.
+//
+// wv and ldwv represent an auxiliary matrix of size nv×(3*nshfts-3).
+func (impl Implementation) Dlaqr5(wantt, wantz bool, kacc22 int, n, ktop, kbot, nshfts int, sr, si []float64, h []float64, ldh int, iloz, ihiz int, z []float64, ldz int, v []float64, ldv int, u []float64, ldu int, nh int, wh []float64, ldwh int, nv int, wv []float64, ldwv int) {
+	checkMatrix(n, n, h, ldh)
+	if !(0 <= ktop && ktop < n) {
+		panic("lapack: invalid value of ktop")
+	}
+	if ktop > 0 && h[ktop*ldh+ktop-1] != 0 {
+		panic("lapack: diagonal block is not isolated")
+	}
+	if !(0 <= kbot && kbot < n) {
+		panic("lapack: invalid value of kbot")
+	}
+	if kbot < n-1 && h[(kbot+1)*ldh+kbot] != 0 {
+		panic("lapack: diagonal block is not isolated")
+	}
+	if nshfts < 0 || nshfts%2 != 0 {
+		panic("lapack: invalid number of shifts")
+	}
+	if len(sr) != nshfts || len(si) != nshfts {
+		panic(badSlice) // TODO(vladimir-ch) Another message?
+	}
+	if wantz {
+		if !(0 <= iloz && iloz <= ihiz && ihiz < n) {
+			panic("lapack: invalid value of iloz or ihiz")
+		}
+		checkMatrix(n, n, z, ldz)
+	}
+	checkMatrix(nshfts/2, 3, v, ldv) // Transposed w.r.t. lapack.
+	checkMatrix(3*nshfts-3, 3*nshfts-3, u, ldu)
+	checkMatrix(3*nshfts-3, nh, wh, ldwh)
+	checkMatrix(nv, 3*nshfts-3, wv, ldwv)
+	if kacc22 != 0 && kacc22 != 1 && kacc22 != 2 {
+		panic("lapack: invalid value of kacc22")
+	}
+
+	// If there are no shifts, then there is nothing to do.
+	if nshfts < 2 {
+		return
+	}
+	// If the active block is empty or 1×1, then there is nothing to do.
+	if ktop >= kbot {
+		return
+	}
+
+	// Shuffle shifts into pairs of real shifts and pairs of complex
+	// conjugate shifts assuming complex conjugate shifts are already
+	// adjacent to one another.
+	for i := 0; i < nshfts-2; i += 2 {
+		if si[i] == -si[i+1] {
+			continue
+		}
+		swap := sr[i]
```

Comment:
I hope I haven't made a mistake but it seems that both versions produce exactly the same assembly. I'll go with the swap idiom.